internal/controllers/addon/upgradepolicy_status.go

```diff
@@ -59,50 +27,83 @@ func (r *AddonReconciler) handleUpgradePolicyStatusReporting(
 		//
 		// All Addons will be requeued when the client becomes available for the first time.
 		log.Info("delaying Addon status reporting to UpgradePolicy endpoint until OCM client is initialized")
+
 		return nil
 	}
 
-	if addon.Status.UpgradePolicy == nil ||
-		addon.Status.UpgradePolicy.ID != addon.Spec.UpgradePolicy.ID {
-		// The current upgrade policy never received a status update.
-		// Tell them: "we are working on it"
-		err := r.handlePatchUpgradePolicy(ctx, ocm.UpgradePolicyPatchRequest{
-			ID:          addon.Spec.UpgradePolicy.ID,
-			Value:       ocm.UpgradePolicyValueStarted,
-			Description: "Upgrading addon.",
-		})
-		if err != nil {
-			return fmt.Errorf("patching UpgradePolicy endpoint: %w", err)
-		}
-
-		addon.Status.UpgradePolicy = &addonsv1alpha1.AddonUpgradePolicyStatus{
-			ID:                 addon.Spec.UpgradePolicy.ID,
-			Value:              addonsv1alpha1.AddonUpgradePolicyValueStarted,
-			ObservedGeneration: addon.Generation,
-		}
+	if addon.Spec.UpgradePolicy == nil || addon.UpgradeCompleteForCurrentVersion() {
 		return nil
 	}
+	if addon.Status.UpgradePolicy == nil || addon.Status.UpgradePolicy.Version != addon.Spec.Version {
+		return r.reportUpgradeStarted(ctx, addon)
+	}
```

Comment:
@Ajpantuso I was thinking about this and the impact to the existing fleet...

For an existing installation, we will  enter this if because of `addon.Status.UpgradePolicy.Version != addon.Spec.Version`, then we will run  `r.reportUpgradeStarted(ctx, addon)` to a policy which might be either in `scheduled` (and never reported by ADO because of the issue here) or in `completed`, so the `reportUpgradeStarted` will keep failing.

In order to avoid that situation, we could be lenient with failures on `reportUpgradeStarted` and `reportUpgradeCompleted` when calling `handlePatchUpgradePolicy`, or, better, check the current value in the endpoint to decide what to do.

Reply:
Either way, we can deprecate the mechanism in the next release, as all CRs status will be compliant with the new format.

Reply:
Good catch, we could also add a step to populate the `status.upgradePolicy.Version` if empty before hitting these conditionals. That will never be the case going forward, but in this transition period that will bring existing CR's in to compliance.

Reply:
I guess the trade off is that would prevent automatic fix of stuck upgrades in which case like you suggested we need to inspect the current UpgradePolicy state first.